Intro to open source with knative blog series p0 and p1 (#5618)
knative/docs: blog/docs/articles/getting-started-blog-p1.md

**Date: 2023-07-06**

Welcome back to this introductory blog series! In this article we are going to be providing an introduction to open source: what it is, why you should care, and how you can participate.

If you already have lots of open source experience, you should be able to completely skip this poast and jump to the next one, where we will go over how to set up your development
environment for working on __Knative__. But if you're new to open source, interested in learning more, or want a refresher then we look forward to discussing the what, why, and 
how of open source with you!

## What is Open Source?

So, what is open source? The key concept behind open source is taht all of the source code is publicly available: anyone can read it, change it, improve on it, and distribute it 
as they see fit [[1]](https://www.redhat.com/en/topics/open-source/what-is-open-source).

Similar to open source software, there is also free software, which differs primarily from open source by having a less commercial focus [[2]](https://writefreesoftware.org/learn).

> While there are important philosophical differences between open source software and free software, we are not going to go into those in this article as this is intended as a more

Author: I think that since we aren't capitalizing or bolding open source software, we also shouldn't capitalize or bold free software. WDYT @Leo6Leo ?

suggestion may not be feasible. However, unless such circumstances exist, make an effort to incorporate all their proposed changes. This approach will foster a productive 
and collaborative code review process.

Another thing to keep in mind is that many projects will have __automated tests__, which will run on your PR to make sure that everything is still working after your changes.
If the tests seem to be failing, try and investigate why and see if you can fix them. If you are unsure why a test is failing, you can always leave a commend on your PR indicating
that and ask for help. Sometimes, tests can behave inconsistently, and are therefore called "flaky" tests. If you are confident that the test failures are not a result of your
code changes, it may be worth re-running the tests.

Once the tests on your PR have passed and your changes have been approved by the reviewer(s), your code will be merged into the main branch upstream by a project maintainer.
Congratulations! You just contributed code to the project and made an impact on the project and community!

> ## Conclusion

Reviewer: At the end, can you create a link to https://contribute.cncf.io/contributors/getting-started/ for people who wants to have a more generic (non-Knative) introduction to open source?

Author: Just added it!

1. Fork the project and clone it locally

2. Make a branch

3. Commit

4. Create a PR

5. Request reviews

6. Make changes upon request

7. PR is approved.

8. Your code gets merged!

Firstly, you should generally be working on a fork of the project. You can think of a fork as your own personal copy of the repository, where you have the permissions to make whatever
changes you want. This is different from the main project repository, where you likely don't have permissions to change the code. To create a fork on GitHub, you can follow 
[these instructions](https://docs.github.com/en/get-started/quickstart/fork-a-repo){:target="_blank"}. To create a fork on GitLab, you can follow [these instructions](https://docs.gitlab.com/ee/user/project/repository/forking_workflow.html){:target="_blank"}.
Some important terminology here is that the repository you forked is generally referred to as the "upstream" repository, and we will be referring to your fork as the "origin",
as that is normally how your local git remotes will be set up. To easily clone a forked repository and set up the git remotes properly, you can use the `git clonefork` command

> which can be installed from [here](https://github.com/cali0707/git-utils){:target="_blank"}. Alternatively, you can use the [git clone](https://git-scm.com/docs/git-clone){:target="_blank"} and 

Reviewer 2: Maybe add a quick code example for cloning the fork and adding a remote to it

[git remote](https://git-scm.com/docs/git-remote){:target="_blank"} commands yourself to configure your local copy of your fork. Note that you will want to add a remote to the upstream repository,

> and that this remote is commonly named "upstream".

Reviewer 2: `Note that you will want to add a remote to the upstream repository, and that this remote is commonly named "upstream".`
Not sure what you mean with that?

Author: I added a bit of explanation for this

you wrote from the rest of the code, then you have done a good job following the project's style. A useful list of things to keep in mind as you work on your code is:

1. Variable, function, and class naming conventions

2. Code formatting/automatic formatter the project is using

3. Is there a commit message convention?

4. File naming conventions

5. Package import rules

6. Package installment policy (check if there is one or if you can install whatever you need)

When you feel like you have made some progress or that your changes are done, be sure to make a commit. For more information on making commits, read [here](https://www.atlassian.com/git/tutorials/saving-changes/git-commit){:target="_blank"}.
Git commits need a message, so try and provide something as this will make it easier for others to understand your changes. For example, "worked on bug fix" __would not be a great
commit message__, while "added try catch block to prevent crash" would be better as it __describes what you changed__.

### Making a PR/MR

![](/blog/images/getting-started-blog-series/post1/ready-for-review.png)

> There are two scenarios where you would likely want to open a PR (or a Merge Request (MR) if your project uses GitLab). The first is that you feel your changes are complete and 

Reviewer 2: ```suggestion
There are two scenarios where you would likely want to open a PR/MR. The first is that you feel your changes are complete and 
```

are ready to be merged into the project. The second is that you have made some progress but are not finished yet, and would like to get some early feedback or assistance on your changes.

> In both cases, you will want to go through the basic flow to create a PR or MR. If you are using GitHub and are unsure of how to do this, you can follow the instructions

Reviewer 2: ```suggestion
In both cases, you will want to go through the basic flow to create a PR/MR. If you are using GitHub and are unsure of how to do this, you can follow the instructions
```

If you are using GitLab and are unsure of how to do this, you can follow the instructions [here](https://docs.gitlab.com/ee/user/project/repository/forking_workflow.html#merge-changes-back-upstream){:target="_blank"}.

When creating your PR/MR it is important to provide the context of why you are making the changes, what you changes, and how people can test your changes. It is good practice to 
include a link to the issue the PR/MR is addressing, and to give a bried description of the changes you made. Additionally, your PR/MR should have a brief and descriptive title
to make it clearer to reviewers what it contains.

If your changes are incomplete, you will want to mark the PR/MR as a __draft__, to indicate to reviewers that your work is not complete. Instructions for Github are 
[here](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request#converting-a-pull-request-to-a-draft){:target="_blank"}, and instructions for Gitlab are [here](https://docs.gitlab.com/ee/user/project/merge_requests/drafts.html#mark-merge-requests-as-drafts){:target="_blank"}. You may also

> want to title your PR/MR as "[WIP]: \<descriptive title>". If you have any questions about your changes, this draft PR/MR is a great to command asn ask them as whoever answers

Reviewer 2: ```suggestion
want to title your PR/MR as "[WIP]: \<descriptive title>" (with _WIP_ for "work in progress"). If you have any questions about your changes, this draft PR/MR is a great to command and asn them as whoever answers
```

> you will have the context of everything you have done so far. Once you feel your changes are complete/ready, you can convert your PR/MR from its draft status to being ready.

Reviewer 2: `If you have any questions about your changes, this draft PR/MR is a great to command asn ask them as whoever answers you will have the context of everything you have done so far.`
Not sure, what you want to say here

Reviewer 2: Hi @Leo6Leo,
thanks for your reply. My issue is mostly with the wording this part `... this draft PR/MR is a great to command asn ask them as whoever answers...` (or does "asn" stand for something specific?)

process, improving your code until it is ready to be merged into the project. Let's discuss this review process next.

### PR/MR Review Process

![](/blog/images/getting-started-blog-series/post1/getting-feedback.png)

Once you have a PR/MR which is ready, one or more people will review your changes and provide feedback. If this is one of your first contributions to the project, expect lots of
feedback! It is then up to you to make changes to address those comments. To do this, all you need to do is make a new commit on the same branch you made the PR from, and push the
changes to your origin.

While it may be frustrating when reviewers suggest changes to what you perceive as functioning code, remember their goal is to enhance the overall quality of the project. 
They provide their feedback with the intention of helping you refine and advance your coding skills! Instead of taking the feedback personally, try and look at it as a learning
opportunity. If there's a reason preventing you from implementing a requested change, don't hesitate to respond to their comment explaining why you believe the 

> suggestion may not be feasible. However, unless such circumstances exist, make an effort to incorporate all their proposed changes. This approach will foster a productive 

Reviewer 2: Great! Really important :+1:
